docs/GOVERNANCE.md

```diff
@@ -0,0 +1,130 @@
+  requests; or
+- reviewing
+  [open pull requests](https://github.com/napari/napari/pulls),
+
+among other possibilities. Any community member can become a contributor, and
+all are encouraged to do so. By contributing to the project, community members
+can directly help to shape its future.
+
+Contributors are encouraged to read the [contributing guide](CONTRIBUTING.md).
+
+## Core developers
+
+Core developers are community members that have demonstrated continued
+commitment to the project through ongoing contributions. They
+have shown they can be trusted to maintain napari with care. Becoming a
+core developer allows contributors to merge approved pull requests, cast votes
+for and against merging a pull-request, and be involved in deciding major
+changes to the API, and thereby more easily carry on with their project related
+activities. Core developers appear as organization members on the napari
+[GitHub organization](https://github.com/orgs/napari/people). Core
+developers are expected to review code contributions while adhering to the
+[core developer guide](CORE_DEV_GUIDE.md).
+
+New core developers can be nominated by any existing core developer.
+Discussion about new core developer nominations is one of the few activities
+that takes place on the project's private management list. The decision to
+invite a new core developer must be made by “lazy consensus”, meaning unanimous
+agreement by all responding existing core developers. Invitation must take
+place at least one week after initial nomination, to allow existing members
+time to voice any objections.
+
+## Steering Council
+
+The Steering Council (SC) members are core developers who have additional
+responsibilities to ensure the smooth running of the project. SC members are
+expected to participate in strategic planning, approve changes to the
+governance model, and make decisions about funding granted to the project
+itself. (Funding to community members is theirs to pursue and manage.) The
+purpose of the SC is to ensure smooth progress from the big-picture
+perspective. Changes that impact the full project require analysis informed by
+long experience with both the project and the larger ecosystem. When the core
+developer community (including the SC members) fails to reach such a consensus
+in a reasonable timeframe, the SC is the entity that resolves the issue.
+
+The steering council is fixed in size to XXXXX members. This may be expanded in
+the future. The initial steering council of napari consists of XXXXXXXXX.
+The SC membership is revisited every January. SC members who do
+not actively engage with the SC duties are expected to resign. New members are
+added by nomination by a core developer. Nominees should have demonstrated
+long-term, continued commitment to the project and its [values](VALUES.md). A
+nomination will result in discussion that cannot take more than a month and
+then admission to the SC by consensus.
```

Comment:
Personally I would like Nick, Loic, and myself to form the initial steering committee, with a view to growing it to 5 people within ~2 years, then keeping it at that level for the foreseeable future. I would also replace the SKIPs with steering committee consensus-seeking followed by a vote. The SC would act as a "multi-BDFL", and we have final authority on project direction, values, governance etc. Eventually we can move from BDFL to a SKIP-like process where all core members vote, but I think that we are still early enough in the project's lifetime that the BDFL model is most useful.

To elaborate a bit more, I think projects need to reach a certain level of maturity and stability before becoming "community-led". I don't want napari to suddenly head off in some random direction, for example, that best suits GIS applications at the expense of microscopy, because the GIS community is much bigger in the scientific Python space. Instead, *as I see it*, napari's main goal remains to fulfil my and especially @royerloic's vision and needs for a microscopy viewer, processor, and maybe one day controller. Contributions to make it e.g. more GIS friendly should be weighted against whether they interfere with that need.

Now, I don't want to sound harsh and I do mean the B in BDFL. ;) But this is my view of governance right now, but I am open to discussion, suggestion, overruling, etc. Just wanted to lay it out there. =)

Reply:
I am open to any initial SC formation, but to be honest what you are saying here about multi-BDFL model contradicting your state so far on **open-source napari** and its community @jni (sorry if this was a too dramatic expression :) ). I cannot see what is **really** the direction of such possible multi-BDFL SC? I think no matter the formation of SC, the direction of SC should be cleared. In an ideal case, direction should be cleared by community, or at least accepted/confirmed by community. Otherwise contributors would ask the natural question of where is this going. I, personally, cannot expect people to invest time and effort on an open-source project without seeing the direction.

Reply:
Open source and governance are orthogonal concerns. The Python language itself has always been open source but, until last year, Guido was BDFL, and he had the final say on what went in and what stayed out. Similarly, although it was unsaid, Gaël Varoquaux was de-facto BDFL for scikit-learn and Stéfan van der Walt was BDFL for scikit-image, until those projects got a community-based governance model.

This is not to say that the source was closed, or that outsiders couldn't contribute. Outsiders *did* contribute to all those projects, a lot. But you had to convince the BDFLs that the contribution was worthwhile. Generally, if you have a minimum amount of trust in the competence of the SC, then this is not a big hurdle for actually valuable contributions. =)

I agree with you that the direction and vision should be clear, so that people can decide whether to invest. That is the purpose of the values document. Sorry that I haven't read it in detail yet, but, napari *does* have a vision and a set of founding values, which I tried to lay out in the blog post, but maybe didn't do as good a job as necessary. =) We also have public roadmaps e.g. in #420. And we have a few open forums (that we are aiming to increase) where people can ask such questions.

So, all this to say, we should help people to decide whether to invest, yes, but a BDFL model does not contradict open source, and if there is a change that Nick, Loic, and I all agree shouldn't be merged, then I *absolutely* think there should be no way that thing gets merged, until someone can convince at least two of us that it should. (Sorry if *that* is too dramatic. =P)

Reply:
let's take a contributor who offered something on one public roadmap like #420  and didn't get approval from SC. Isn't such situation is the thing push people forking and separating ways? I just want to avoid that. I believe values (i personally trust any SC about values, that's why happy with the everybody involved so far) should be approved by community and direction can come from a multi-BDFL, but a strict governance on direction will put people off easily. Something to be cautious about. 

 **Also**, what if there will be an unresolved issue within SC about direction later on? No mechanism to handle those :)

Reply:
> let's take a contributor who offered something on one public roadmap like #420 and didn't get approval from SC. Isn't such situation is the thing push people forking and separating ways? I just want to avoid that.

Yes, we should try to avoid forking, of course! But, it is always a risk, and you have to believe that Nick, Loic, and I won't just dismiss contributions/feature requests out of hand. I think we have a reasonable track record on this. =) If we, together, reject a proposal after discussion, I believe that the proposal *should* be rejected.

> what if there will be an unresolved issue within SC about direction later on?

The SC should operate on consensus-seeking followed by a vote. No decision should just be voted on without prior discussion and deliberation, but if there is still disagreement after that, it makes sense that Loic and Nick should be able to overrule me, Loic and I should be able to overrule Nick, and (I am least sure of this one, but happy to stand by it in defence of symmetry =P) Nick and I should be able to overrule Loic.

Reply:
@AhmetCanSolak strict governance is essential for every open source project that wants to establish a sustainable community, including at some point potentially raising and spend its own funding.

It is also standard within the scientific python ecosystem - for example [numpy has a steering counci](https://docs.scipy.org/doc/numpy-1.15.1/dev/governance/governance.html#steering-council), [scipy has a Benevolent Dictator For Life (BDFL)](https://docs.scipy.org/doc/scipy/reference/dev/governance/governance.html#bdfl) in addition to a steering council, and [scikit-learn has a technical committee](https://scikit-learn.org/stable/governance.html). [jupyter has a BDFL + steering council](https://github.com/jupyter/governance/blob/master/people.md). [matplotlib also as a BDFL + steering council](https://github.com/matplotlib/governance/blob/master/people.md). Collectively these projects have supported 1000s of contributors with their decision making process.

Does what we are proposing here for napari differ from your interpretations of what is true for these projects (or others in the scientific python ecosystem)? Or do you also have specific concerns about how those projects are currently governed that you want napari to learn from as we frame our governance process?

Reply:
@sofroniewn we can definitely learn from others. I would say such more strict governance will make more sense in the long run. I believe napari is on an early stage still. We are getting new great contributors constantly and i am grateful for everyone contributed so far. Just concerned about cutting this nice flow with our initial governance regulation.

>  If we, together, reject a proposal after discussion, I believe that the proposal _should_ be rejected.
> 

@jni 
I exactly want to avoid such sentences for possible future contributors here :) It should be written that with enough/significant community support, a contributor can make a decision whether SC agrees or not. napari is still on an early stage and my stand is keeping written governance less strict at this stage. I am okay with whatever the final consensus of people here. Just putting out there what i think.

Reply:
@AhmetCanSolak these are good points, and now is definitely the time to raise them.

Times of irresolvable disagreement between the community, contributors, core developers, or the steering council will always be hard for an open source project, and in those situations the governance model we're looking to adopt asks us to try and find consensus first, but ultimately there must be a way to decide and looking at those other successful python projects having combinations of BDFL and steering councils seems to have worked well for them.

my hope is that by having a governance model it will actually make the project more welcoming for new contributors as they will explicitly understand how decisions are made, and how they can grow to have more of a voice in the decision making process, for example becoming a core dev, or ultimately be on the steering council themselves. we need to make sure the document makes these things clear though

Reply:
> I would say such more strict governance will make more sense in the long run. I believe napari is on an early stage still.

I think you are misunderstanding the history and purpose of such documents. Yes, almost none of them existed originally, but what they do, in general, is *lay out existing processes*, rather than impose new processes. Matplotlib didn't have a governance doc early on, but they had a BDFL, first in John Hunter, then Michael Droetboom, then Tom Caswell. Where they do impose new processes, those processes are *more liberal*, not less. See for example [this tweet](https://twitter.com/GaelVaroquaux/status/1103745426568724482) by Gaël Varoquaux: a contribution that had been blocked by him for years was finally able to get in. Yes, he makes the argument that this is how it should be, but that's my point: this happened when scikit-learn was a very mature library, and everyone agreed on what it was supposed to do, what made it valuable, etc. At an early stage, there are still many possible futures, and you need someone to trim those down.

Indeed, what makes scikit-learn valuable is it's rock-solid, super-consistent API, and getting there required a lot of "saying no" from Gaël. 

So, I agree with you that we are in an early stage, but imho that means we need *more* direction and vision, not less, just as those projects had. When you get a giant influx of contributions early, two things can happen:
- the software grows and evolves rapidly, but
- the software becomes a complex beast that crumbles under its own weight.

You need clear guidance to avert the second fate. This is the reason why we waited so long to announce napari, btw, and the same principle applies now.

> I exactly want to avoid such sentences for possible future contributors here :)

Oh, 100%! Going back to the earlier point about the purpose of a governance document: it is to be *welcoming* to new contributors, not onerous or scary. (Echoing @sofroniewn's point.) The idea is for a new potential contributor to the repo to be able to quickly see: ok, if I want to contribute something, who do I need to discuss it with? If some handle I don't recognise on GitHub says, yep this would be a valuable contribution, is this an authoritative voice, or just a supportive one? The converse is also true and very important. We actually had a case in scikit-image where Thomas Walter (now a core dev :sweat_smile:) asked, would this contribution be welcome, and some random from the internet said, why don't you just fork the repo, and Thomas said Ok good point, and closed the issue! For some reason, both Stéfan and I missed the whole exchange, and it's only because Nelle pointed it out to us that we were able to correct the situation and get him to make his contribution!

At any rate, I'll second @sofroniewn's comment also: the concern that we should *welcome* new contributors and not drive them away is a major one, and we need to word the governance document accordingly. But, welcoming contributions and welcoming any and all contributions is not the same thing.

Reply:
My opinion on this: I feel strongly in avoiding that the original design principles of Napari be scattered and confused by all passing tastes and ideas. So I think the SC is necessary (Nick, Juan and myself) to give some firm direction to the boat. Down the lines, things will change. But, first, let's stay true to our core principles and attract more like minded contributors so that eventually, we have people that will perpetuate the core ideas behind napari.
I am not worried about forking. I am confident that we will reach good consensus decisions that will make forking unnecessary and futile...

Reply:
> eventually, we have people that will perpetuate the core ideas behind napari.

👆 that thing!